### PR TITLE
Loops

### PR DESCRIPTION
**Pull request by orius123**
compiler is handled

### CODE REVIEW

**Comment by orius123** [comment]
works.
not enough tests.
no break functionality.

**Comment by meirwah** [comment]
@Bonczidai  please review and see that your Deis story (looping on CoreOs cluster), works with this soultion/feature

**Comment by orius123** [comment]
ready

File: score-lang-compiler/src/main/java/org/openscore/lang/compiler/modeller/ExecutableBuilder.java [hunk 1]
```diff
@@ -183,24 +199,29 @@ private Workflow compileWorkFlow(LinkedHashMap<String, Map<String, Object>> work
...
         String defaultFailure = isOnFailureDefined ?
-                onFailureWorkFlow.getTasks().getFirst().getName() : ScoreLangConstants.FAILURE_RESULT;
+                onFailureWorkFlow.getTasks().getFirst().getName() : FAILURE_RESULT;
 
         while (iterator.hasNext()) {
             Map.Entry<String, Map<String, Object>> taskRawData = iterator.next();
             Map.Entry<String, Map<String, Object>> nextTaskData = iterator.peek();
             String taskName = taskRawData.getKey();
             Map<String, Object> taskRawDataValue;
+            String message = "Task: " + taskName + " syntax is illegal.\nBelow task name, there should be a map of values in the format:\ndo:\n\top_name:";
```
Comment (meirwah) [inline score-lang-compiler/src/main/java/org/openscore/lang/compiler/modeller/ExecutableBuilder.java]: hoyld this be a string const at Class level?

File: score-lang-compiler/src/main/java/org/openscore/lang/compiler/modeller/transformers/ForTransformer.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+@Component
+public class ForTransformer implements Transformer<String, ForLoopStatement>{
+    @Override
+    public ForLoopStatement transform(String rawData) {
+        if (StringUtils.isEmpty(rawData)) {
+            return null;
+        }
+        String inKeyword = " in ";
+        String regex = "^(\\s+)?(\\w+)\\s+(in)\\s+(\\w+)(\\s+)?$";
```
Comment (meirwah) [inline score-lang-compiler/src/main/java/org/openscore/lang/compiler/modeller/transformers/ForTransformer.java]: const at class level?

**Comment by Bonczidai** [comment]
I tested the looping mechanism with a complex flow (iterated over a list of machines and cleared the docker images) - it works.

Question:
what is the stage of the aggregate outputs (publish) feature?
It seems to me that this syntax doesn't update the agg_result variable:
```
print_hosts:
  loop:
    for: host in hosts
    do:
      ops.print:
        - host
        - port: default_port
    publish:
      - agg_result: agg_result + op_result
```

**Comment by orius123** [comment]
You are right. I opened an issue for it. If it blocks you let me know an I will prioritise it.

**Comment by Bonczidai** [comment]
Turns out I need the aggregate feature for my use cases..

**Comment by orius123** [comment]
Ok. Will move it up.

**Comment by orius123** [comment]
fixed and merged back from master.
@Bonczidai can you please try again, so we will be  able to merge it?

**Comment by Bonczidai** [comment]
It seems to me that it does not work :( . Here is what I observed:

using syntax:
```
            for: ...
            do: ...
            publish:
                - agg_result: >
                    op_output # (1)
                    #agg_result + op_output  # (2)
                    #agg_result + op_output if agg_result != None else ""  # (3)
```

+ agg_result is evaluated only after the last iteration (in previous iterations *EVENT_OUTPUT_START* is followed immediately by a *EVENT_INPUT_END* not *EVENT_OUTPUT_END*)
+ if you use format (2) or (3) the expression is not evaluated since agg_result does not exist in the context; and if you define it before for loop (flow level) it can not be overridden

Any suggestion? Am I doing something wrong?

**Comment by orius123** [comment]
Why when you define it as flow input it can't be overriden?

**Comment by Bonczidai** [comment]
I am not sure, see:

case1:
```
  inputs:
    - agg_result:
        default: "'INIT'"
        
    check:
        loop:
            for: docker_host in docker_host_list
            do:
                l.print:
                    - text: docker_host
            publish:
                - agg_result: >
                    agg_result + op_output
```
agg_result = "INIT"

case2:
```
  inputs:
    - agg_result:
        default: "'INIT'"
        
    check:
        loop:
            for: docker_host in docker_host_list
            do:
                l.print:
                    - text: docker_host
            publish:
                - agg_result: >
                    op_output
```
agg_result = op_ouput from third iteration

case3:
```
 # agg_result undefined
        
    check:
        loop:
            for: docker_host in docker_host_list
            do:
                l.print:
                    - text: docker_host
            publish:
                - agg_result: >
                    agg_result + op_output
```
error

**Comment by orius123** [comment]
Assuming that op_output is a string and no casting is required,
You need to add the `fromInputs` keyword in the expression. i.e:
```yaml
publish:
  - agg_result: fromInputs['agg_result'] + op_output
```

**Comment by Bonczidai** [comment]
Great.. it works with this syntax if `agg_result` is defined as a flow input

**Comment by orius123** [comment]
@Bonczidai if you find the pulll request fine. please merge it back to master.

**Comment by Bonczidai** [comment]
Could you resolve conflicting files?

(I can also try to merge them, just let me know)

**Comment by orius123** [comment]
I'm out of office today so I can do it on wednesday, if there are no conflicts, I would thank if you do it.

**Comment by Bonczidai** [comment]
resolved four conflicting classes and fixed testing issues..
merging into master

**Pull request merged**